src/lib/sm2.c

```diff
@@ -0,0 +1,271 @@
+/*-
+ * Copyright (c) 2017 Ribose Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <botan/ffi.h>
+
+#include "sm2.h"
+#include "crypto.h"
+#include "packet.h"
+#include "readerwriter.h"
+#include "utils.h"
+#include "utils.h"
+
+extern ec_curve_desc_t ec_curves[PGP_CURVE_MAX];
+
+pgp_errcode_t
+pgp_sm2_genkeypair(pgp_seckey_t *seckey, pgp_curve_t curve)
+{
+    /**
+     * Keeps "0x04 || x || y"
+     * This follows the same encoding as ECDSA
+     *
+     * P-521 is biggest supported curve for SM2, though ordinarilly
+     * it uses the 256-bit SM2 recommended curve.
+     */
+    uint8_t         point_bytes[BITS_TO_BYTES(521) * 2 + 1] = {0};
+    const size_t    filed_byte_size = BITS_TO_BYTES(ec_curves[curve].bitlen);
+    botan_privkey_t pr_key = NULL;
+    botan_pubkey_t  pu_key = NULL;
+    botan_rng_t     rng = NULL;
+    BIGNUM *        public_x = NULL;
+    BIGNUM *        public_y = NULL;
+    pgp_errcode_t   ret = PGP_E_C_KEY_GENERATION_FAILED;
+
+    if (botan_rng_init(&rng, NULL)) {
+        goto end;
+    }
+
+    if (botan_privkey_create(&pr_key, "SM2_Sig", ec_curves[curve].botan_name, rng)) {
+        goto end;
+    }
+
+    if (botan_privkey_export_pubkey(&pu_key, pr_key)) {
+        goto end;
+    }
+
+    // Crash if seckey is null. It's clean and easy to debug design
```

Comment:
I feel like we ought to start adding some debug assertions for things like this, rather than crashing. But that's a conversation for another day.